### PR TITLE
Tempo: Add start time and end time parameters while querying traces

### PR DESCRIPTION
**Pull request by bikashmishra100**
**What this PR does / why we need it**:
Recently https://github.com/grafana/tempo added start and end timestamp to /api/traces API via https://github.com/grafana/tempo/pull/1388. This is done to make the searches faster  by scanning only the blocks within the start and end time. We want to consume the parameters so that the start and end timestamps are supplied from Grafana.

**Special notes for your reviewer**:
Aligning this PR with @joe-elliott. I hope this PR satisfies the requirements that was mentioned in https://github.com/grafana/tempo/issues/1373. The timestamps will be the range of the search query given by the user.

cc: @sagarwala, @ashwinidulams

### CODE REVIEW

**Comment by joe-elliott** [comment]
It seems like the time range is based on the same as the search. Is this true? 

My concern with this is that you can find a trace using search, but not receive the entire trace when you query by id if we make this change. If the trace overlaps either the start or end of the search range it is possible it will not be returned in its entirety. this is unlikely, but possible.

I'm wondering if we should make this configurable, defaulted to false for backwards compat?

**Comment by bikashmishra100** [comment]
Sure, making it configurable is what I can work on.

I too had this thought that if user provides a very narrow window to search for logs in case the exact time of an error is known, the entire trace may or may not fit inside the window. Unfortunately I do not have any great ideas at this moment. Does fixed buffers (which also could be configurable) e.g. 30 minutes beyond the search window make sense ?

**Comment by joe-elliott** [comment]
I think a hardcoded value like `30m` would be a fine place to start, but likely this should be configurable as well. Thanks for looking into this improvement!

File: pkg/tsdb/tempo/tempo_test.go [hunk 1]
```diff
@@ -10,9 +10,16 @@ import (
...
+		req, err := service.createRequest(context.Background(), &datasourceInfo{}, "traceID", 0, 0)
+		require.NoError(t, err)
+		assert.Equal(t, 1, len(req.Header))
+	})
+
+	t.Run("createRequest with time range - success", func(t *testing.T) {
+		service := &Service{tlog: log.New("tempo-test")}
+		req, err := service.createRequest(context.Background(), &datasourceInfo{}, "traceID", 1, 2)
 		require.NoError(t, err)
 		assert.Equal(t, 1, len(req.Header))
```
Comment (connorlindsey) [inline pkg/tsdb/tempo/tempo_test.go:24]: Would be nice to also assert that the query is formatted correctly in this test case

**Review by connorlindsey — CHANGES_REQUESTED** [review]
Looks good, but needs to take into account a configurable time shift

**Comment by bikashmishra100** [comment]
Apologies for not being able to work on this recently as I had to work on and mitigate a production incident. I will work on this over the weekend and update it.

**Comment by connorlindsey** [comment]
@bikashmishra100 Absolutely no worries, we really appreciate you being willing to open a PR for this. If you need any help or want to pass this off for us to finish up, we can help too

**Comment by bikashmishra100** [comment]
No problem, this is a trivial thing. I would not bother you/your team. I will work on it and update here.

**Comment by hanjm** [comment]
still useful

**Comment by hanjm** [comment]
@connorlindsey  @bikashmishra100 I think it could use the span start time with some duration range instead of the search time, like spanStartTime-1h~spanEndtime+1h. 

i implement a tricky proxy in production, it add timestamp in traceid string (like xxxx_ts), then when we click to view trace detail,  the proxy split special traceid to traceid and timestamp, which make the performance improve fast a lot!

**Review by aocenas — APPROVED** [review]

File: public/app/plugins/datasource/tempo/configuration/QuerySettings.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  return (
+    <div className={styles.container}>
+      <h3 className="page-heading">TraceID Query</h3>
+      <InlineFieldRow>
+        <InlineField
+          label="Span start time shift"
+          labelWidth={26}
+          grow
+          tooltip="Shifts the start time of the span. Default 0 (Time units can be used here, for example: 5s, 1m, 3h)"
```
Comment (aocenas) [inline public/app/plugins/datasource/tempo/configuration/QuerySettings.tsx:22]: I think this needs a bit longer explanation and better name.

label: Time shift for start of search time range
tooltip: Shifts the start of the time range when searching by trace ID. This is needed as searching for traces can return traces that do not fully fall into the search time range. If trace is searched by trace id in the same time range it won't show up. This shift should allow for traces to fall fully into the shifter time range. The longer your traces the higher the time shift should be used. Default 30m (Time units can be used here, for example: 5s, 1m, 3h).

@joe-elliott does this make sense? Also I wonder do we need start shift? Don't we only need the end shift? Meaning does the search return traces that started before the search time range but overlap a bit? Cause I would assume we need only end range shift.

**Comment by aocenas** [comment]
>If the trace overlaps either the start or end of the search range it is possible it will not be returned in its entirety. this is unlikely, but possible.

@joe-elliott to understand this: Does this mean that if you query by trace id and there is partial overlap with the trace start-end and the time range, tempo will return partial trace? So at this moment even if we use the shift, this can still happen and user won't really know that the trace is partial, is that right?

**Comment by empollard** [comment]
Hi Team! I see a label on this for no-backport. We currently have  a customer waiting for this fix and it looks like we don't have an ETA for 9.3.0 yet. Would there be some way to backport this to 9.2 or 9.1.6?

Thanks!

**Review by aocenas — APPROVED** [review]

**Review by connorlindsey — APPROVED** [review]

**Comment by joe-elliott** [comment]
> @joe-elliott to understand this: Does this mean that if you query by trace id and there is partial overlap with the trace start-end and the time range, tempo will return partial trace? So at this moment even if we use the shift, this can still happen and user won't really know that the trace is partial, is that right?

@aocenas That is correct. When specifying the start/end parameters you are instructing Tempo to search a subset of the blocks on the backend. This may result in a partial trace being returned. It will be up the operator to set the start/end time shift parameters to feel confident that all spans for a given trace will be within the window.

**Comment by aocenas** [comment]
@joe-elliott is there some metadata in the response indicating that this is happening? So that we can show some warning in the UI something like "showing partial trace, widen time range to see it whole".

**Comment by joe-elliott** [comment]
There is no way to know if the trace exists elsewhere except for Tempo to search for it. At which point it's basically performing the full search anyway.

This feature is for those people who have a long retention at a high volume. In this case you could set the start/end time shift to something like `1d` and vastly reduce the amount of data searched when doing a trace by id lookup. if your traces are normally only a few seconds you can feel quite comfortable that you are getting nice performance gains with an extremely low chance of returning a partial trace.

**Pull request merged**